### PR TITLE
Jetpack Social: Prevent SIG from enabling itself automatically on existing posts

### PR DESCRIPTION
**Pull request by danielpost**
## Proposed changes:

This PR fixes an issue where Social Image Generator was being enabled automatically for existing posts, by relying on Gutenberg to update the post meta rather than doing it a hook after saving the post. I've explained the changes in detail in the code since I think the context is helpful.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

1202816441505085-as-1204236160532606/f

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* Make sure Social Image Generator is enabled for your site.
* Create a new post and make sure the Social Image Generator toggle is enabled.
* Publish it or save it as a draft.
* Check that the meta was saved correctly: `get_post_meta( POST_ID )` in the shell
* Now, delete the post meta (`delete_post_meta( POST_ID, '_wpas_options' )` and open the editor for the post.
* Check that the toggle is disabled.
* Change the post title or content and save it.
* Check the meta again, and make sure that SIG is now disabled.
* Repeat this process for the Classic Editor. In this case, SIG should always be disabled.

### CODE REVIEW

File: projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js
Comment (danielpost) [inline projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js]: When creating a new post or editing an existing one, the [default value for enabling SIG is already set here](https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js#L8). However, this value was not being updated in the post meta. This `useEffect` hook simply updates the meta with the current value when the SIG panel is opened, so that it gets saved properly.

ESLint is being ignored because it wants `setIsEnabled` and `isEnabled` in the dependency array, which causes an infinite loop.

The tradeoff here is that SIG won't be enabled if a post is created programatically, since the post meta won't be present. This however can be fixed fairly easily by adding the post meta when the post is being created.

File: projects/packages/publicize/src/class-publicize-setup.php
Comment (danielpost) [inline projects/packages/publicize/src/class-publicize-setup.php]: **JUST FOR TESTING PURPOSES!**

I will remove this before merging because we will use https://github.com/Automattic/jetpack/pull/29529 instead.

File: projects/packages/publicize/src/social-image-generator/class-setup.php
Comment (danielpost) [inline projects/packages/publicize/src/social-image-generator/class-setup.php:89]: Since we now always have the meta present if SIG is enabled, we can simplify our check and remove the meta update in this hook.

File: projects/packages/publicize/src/class-publicize-base.php
Comment (pablinos) [inline projects/packages/publicize/src/class-publicize-base.php]: This should probably be swapped to using the constant

File: projects/packages/publicize/src/class-publicize-base.php
Comment (pablinos) [inline projects/packages/publicize/src/class-publicize-base.php]: It might be better if this is a method on the class, so it's easier for someone to remove, if needs be.

File: projects/packages/publicize/src/class-publicize-base.php
Comment (pablinos) [inline projects/packages/publicize/src/class-publicize-base.php]: I don't quite understand why the `value` array has a `0` index, like there would be multiple default values, but this does work.

The other thing to check is if a post has a value for `_wpas_options` (because they've shared an image for example) will we still check the defaults for `image_generator_settings`? If not, then we should probably flip the logic, so the default is `false` but we set it to `true` for `auto-draft` posts.

Comment (pablinos) [reply projects/packages/publicize/src/class-publicize-base.php]: Ok, I've realised what's going on here now. When `get_metadata` is called with `$single=false` then the value is expected to be an array, and so we get passed the default value as an array of length 1.

I've reworked this function a bit to cope with both values for `$single` as, between the tests, and the default loading of a post in the block editor, it can be set differently.

**Review by gmjuhasz — DISMISSED** [review]
This tests well for me, nice one!

**Comment by danielpost** [comment]
I can't approve this because I opened the PR but this works as expected!

**Review by pablinos — APPROVED** [review]
Approving on behalf of @danielpost 😄

**Pull request merged**
